### PR TITLE
feat(pubsub): add open telemetry trace support

### PR DESCRIPTION
**Pull request by hongalex**
This PR adds the foundation for tracing of a publisher and subscriber for Pub/Sub. This PR attempts to mimic existing messaging system tracing and follows the semantic conventions [defined here](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.1.0/specification/trace/semantic_conventions/messaging.md). This is a draft and is subject to backwards incompatible changes.

[Here's a sample](https://gist.github.com/hongalex/929383213155a64b30435b1bc42171dd) that demonstrates using this otel enabled library w/ Google Cloud Tracing as the exporter. The output looks something like the following:
image

### CODE REVIEW

File: pubsub/topic.go [hunk 1]
```diff
@@ -522,19 +528,32 @@ type PublishResult = ipubsub.PublishResult
...
+	}
+
 	// Calculate the size of the encoded proto message by accounting
 	// for the length of an individual PubSubMessage and Data/Attributes field.
 	msgSize := proto.Size(&pb.PubsubMessage{
 		Data:        msg.Data,
 		Attributes:  msg.Attributes,
 		OrderingKey: msg.OrderingKey,
 	})
+	span.SetAttributes(semconv.MessagingMessagePayloadSizeBytesKey.Int(msgSize))
```
Comment (quartzmo) [inline pubsub/topic.go:556]: Is this already set in `getSpanAttributes` on line 532, above?

Comment (quartzmo) [reply pubsub/topic.go:556]: I think I understand: After injecting the span into the message attributes, the size of the message needs to be updated.

File: pubsub/trace.go [hunk 2]
```diff
@@ -236,3 +242,61 @@ func withSubscriptionKey(ctx context.Context, subName string) context.Context {
...
+		OrderingKey: msg.OrderingKey,
+	})
+	ss := []trace.SpanStartOption{
+		trace.WithAttributes(
+			semconv.MessagingSystemKey.String("pubsub"),
+			semconv.MessagingDestinationKey.String(topic),
+			semconv.MessagingDestinationKindTopic,
+			semconv.MessagingMessageIDKey.String(msg.ID),
+			semconv.MessagingMessagePayloadSizeBytesKey.Int(msgSize),
+			attribute.String("pubsub.ordering_key", msg.OrderingKey),
```
Comment (quartzmo) [inline pubsub/trace.go:296]: Should this be `"messaging.pubsub.ordering_key"`, per the design doc?

File: pubsub/topic.go [hunk 3]
```diff
@@ -697,6 +728,7 @@ func (t *Topic) publishMessageBundle(ctx context.Context, bms []*bundledMessage)
 			ipubsub.SetPublishResult(bm.res, "", err)
 		} else {
 			ipubsub.SetPublishResult(bm.res, res.MessageIds[i], nil)
+			bm.span.SetAttributes(semconv.MessagingMessageIDKey.String(res.MessageIds[i]))
```
Comment (quartzmo) [inline pubsub/topic.go:731]: Does the "publish RPC" span also need to be updated with the message ID?

**Comment by mbn18** [comment]
Hey, is there any expectation to when this will be supported? thanks

**Comment by mbn18** [comment]
@hongalex , thanks for letting me know.

For the mean while I took your propagator design and implemented it in my code. Will add it here so other can use it till a final solution exist
```
package whatever

import (
	"cloud.google.com/go/pubsub"
	"context"
	"go.opentelemetry.io/otel"
)

const pubsubAttribPrefix = "whatever"

// PubsubMessageCarrier injects and extracts traces from a pubsub.Message.
type PubsubMessageCarrier struct {
	msg *pubsub.Message
}

// NewPubsubMessageCarrier creates a new PubsubMessageCarrier.
func NewPubsubMessageCarrier(msg *pubsub.Message) PubsubMessageCarrier {
	return PubsubMessageCarrier{msg: msg}
}

// Get retrieves a single value for a given key.
func (c PubsubMessageCarrier) Get(key string) string {
	return c.msg.Attributes[pubsubAttribPrefix+"_"+key]
}

// Set sets an attribute.
func (c PubsubMessageCarrier) Set(key, val string) {
	c.msg.Attributes[pubsubAttribPrefix+"_"+key] = val
}

// Keys returns a slice of all keys in the carrier.
func (c PubsubMessageCarrier) Keys() []string {
	i := 0
	out := make([]string, len(c.msg.Attributes))
	for k := range c.msg.Attributes {
		out[i] = k
		i++
	}
	return out
}

func PubSubMessageInjectContext(ctx context.Context, msg *pubsub.Message) {
	otel.GetTextMapPropagator().Inject(ctx, NewPubsubMessageCarrier(msg))
}

func PubSubMessageExtractContext(ctx context.Context, msg *pubsub.Message) context.Context {
	return otel.GetTextMapPropagator().Extract(ctx, NewPubsubMessageCarrier(msg))
}
```

**Comment by Mistic92** [comment]
Hi, what is the current status?

**Comment by panperla** [comment]
Are there any alternative thread? Or it just died and nobody is looking into it?

**Comment by meredithslota** [comment]
We are still tracking this here: https://github.com/googleapis/google-cloud-go/issues/4665

**Pull request closed without merging**